src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -364,11 +376,219 @@ private static bool IsSafeForReordering(BoundExpression expression, RefKind kind
+                }
+            }
+
+            return visitedArgumentsBuilder.ToImmutableAndFree();
+
+            InterpolatedStringHandlerData? setupInterpolationDataIfNeeded(
+                int argumentIndex,
+                ImmutableArray<RefKind> argumentRefKindsOpt,
+                ref bool receiverStored,
+                ref int lastStoredIndex,
+                ref BoundExpression? visitedReceiver,
+                ref ArrayBuilder<LocalSymbol>? locals,
+                ref ArrayBuilder<BoundExpression>? sideEffectsBuilder,
+                ref BitVector positionsAssignedTemps)
+            {
+                var argument = arguments[argumentIndex];
+
+                if (argument is BoundConversion { ConversionKind: ConversionKind.InterpolatedStringHandler, Operand: BoundInterpolatedString operand })
+                {
+                    // Handler conversions are not supported in expression lambdas.
+                    Debug.Assert(!_inExpressionLambda);
+                    var interpolationData = operand.InterpolationData.GetValueOrDefault();
+                    var creation = (BoundObjectCreationExpression)interpolationData.Construction;
+
+                    if (creation.Arguments.Length > (interpolationData.HasTrailingHandlerValidityParameter ? 3 : 2))
+                    {
+                        Debug.Assert(!((BoundConversion)argument).ExplicitCastInCode);
+
+                        // We have an interpolated string handler conversion that needs context from the surrounding arguments. We need to store
+                        // all arguments up to and including the last argument needed by this interpolated string conversion into temps, in order
+                        // to ensure we're keeping lexical ordering of side effects.
+                        if (locals == null)
+                        {
+                            Debug.Assert(sideEffectsBuilder == null);
+                            Debug.Assert(positionsAssignedTemps.IsNull);
+                            Debug.Assert(refKindsBuilder == null);
+
+                            locals = ArrayBuilder<LocalSymbol>.GetInstance();
+                            sideEffectsBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+                            positionsAssignedTemps = BitVector.Create(arguments.Length);
+
+                            if (!argumentRefKindsOpt.IsDefault)
+                            {
+                                refKindsBuilder = ArrayBuilder<RefKind>.GetInstance(argumentRefKindsOpt.Length);
+                                refKindsBuilder.AddRange(argumentRefKindsOpt);
+                            }
+                        }
+                        Debug.Assert(sideEffectsBuilder != null);
+                        Debug.Assert(!positionsAssignedTemps.IsNull);
+
+                        // Find the last argument needed by the interpolation, and store from the last stored index all the way up to that argument
```

Comment:
>Find the last argument needed by the interpolation, and store from the last stored index all the way up to that argument [](http://example.com/codeflow?start=27&length=120)

I don't think this is necessary. I believe we need to store only the arguments that are involved in creation of builders. Those that aren't involved, shouldn't be stored.